### PR TITLE
[AI Chat] Update AI Chat client including using async thunk

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR updates a few items on the client side of AI Chat. Note that @bencodeorg 's [PR](https://github.com/code-dot-org/code-dot-org/pull/53477) that adds the waiting animation was merged to this PR to more easily resolve merge conflicts.

Changes include:
- Updates the `onSubmit` function to a redux async thunk,
- Displays a 'Waiting for response...' message until chat assistant response is received from chatAPI. Once received, the assistant message is displayed - this is @bencodeorg's [work](https://github.com/code-dot-org/code-dot-org/pull/53477) that was merged to this PR.
- Clears the chat message history when a user navigates to a different level.
- Prevents multiple user message submissions (disables button while waiting for chatApi response)

Using an [async thunk](https://redux-toolkit.js.org/api/createAsyncThunk) abstracts the standard recommended approach for handling async request lifecycles. The `submitChatMessage` function accepts the Redux action type `'aichat/submitChatMessage'` and the callback function returns a promise. Based on the status of the promise, I added `extraReducers` to update state value `isWaitingForChatResponse` accordingly.

I also added the `updateChatMessageStatus` reducer which is dispatched in `submitChatMessage` when the ChatApi response is returned and the new user `ChatCompletionMessage`'s status is updated ('OK', 'INAPPROPRIATE', or 'PERSONAL').

Here is a screencast video including updates. Note that when user is waiting for assistant response, waiting animation is displayed and the submit button is disabled and another user message submission is prevent. Also, when the level is changed, the chat history is now being cleared.

https://github.com/code-dot-org/code-dot-org/assets/107423305/ef09642b-db38-4fff-b534-2ffdabb3c57e

Thanks to @sanchitmalhotra126 for pairing with me to implement the async thunk and @bencodeorg for his work in the waiting for response animation.

## Links
[jira](https://codedotorg.atlassian.net/browse/LABS-14)

## Testing story
I tested the change in functionality locally in /allthethings/lessons/47.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/aichat/redux/aichatRedux.ts [hunk 1]
```diff
@@ -1,35 +1,93 @@
...
 
 const initialState: AichatState = {
-  newUserMessage: '',
   chatMessages: initialChatMessages,
   isWaitingForChatResponse: false,
   showWarningModal: true,
 };
 
+// THUNKS
+export const submitChatMessage = createAsyncThunk(
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/redux/aichatRedux.ts:37]: Great comments throughout this function! Can we add a top-level one here as well just as an overview? Probably something like `// submits a chat message to the chat completion endpoint, waits for a response, and updates message state`

File: apps/src/aichat/redux/aichatRedux.ts [hunk 2]
```diff
@@ -1,35 +1,93 @@
...
+  'aichat/submitChatMessage',
+  async (message: string, thunkAPI) => {
+    const state = thunkAPI.getState() as {lab: LabState; aichat: AichatState};
+    const systemPrompt = (state.lab.levelProperties as AichatLevelProperties)
+      ?.systemPrompt;
+    if (systemPrompt === undefined) {
+      throw new Error('systemPrompt is undefined');
+    }
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/redux/aichatRedux.ts:47]: Just a note for later: we may actually want to handle this in AichatView directly, by checking this value whenever `state.lab.levelProperties` updates; whenever the `levelProperties` change, and there is no system prompt, then we can't run the chat app at all and we should probably throw up an error dialog. But this thunk shouldn't need to handle that as it's called on every chat message submission.

File: apps/src/aichat/redux/aichatRedux.ts [hunk 3]
```diff
@@ -42,14 +100,36 @@ const aichatSlice = createSlice({
...
+      }
+    },
+  },
+  extraReducers: builder => {
+    builder.addCase(submitChatMessage.fulfilled, state => {
+      state.isWaitingForChatResponse = false;
+    });
+    builder.addCase(submitChatMessage.rejected, (state, action) => {
+      state.isWaitingForChatResponse = false;
+      console.error(action.error);
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/redux/aichatRedux.ts:129]: Should we go ahead and add an error field to Aichat state? We don't have to actually handle it in the view yet, but for now we could just do something like `state.chatMessageError = true` (doesn't necessarily have to be called `chatMessageError`)

File: apps/src/aichat/views/ChatWorkspace.tsx [hunk 4]
```diff
@@ -27,84 +24,54 @@ const ChatWorkspace: React.FunctionComponent = () => {
...
-      const assistantChatMessage: ChatCompletionMessage = {
-        id: chatApiResponse.id + 1,
-        role: Role.ASSISTANT,
-        status: Status.OK,
-        chatMessageText: chatApiResponse.assistantResponse,
-      };
-      dispatch(addChatMessage(assistantChatMessage));
-    }
-  };
+  // When the level changes, clear the chat message history.
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatWorkspace.tsx:54]: 🎉

**Review by sanchitmalhotra126 — APPROVED** [review]
Looks great! Couple of minor things/questions but nothing blocking. Nice work!

File: apps/src/aichat/redux/aichatRedux.ts (hunk 1 above)
Comment (fisher-alice) [reply apps/src/aichat/redux/aichatRedux.ts:37]: Yes - definitely can add: Thanks!
```
// This thunk's callback function submits a user chat message to the chat completion endpoint,
// waits for a chat completion response, and updates the user message state.
```

File: apps/src/aichat/redux/aichatRedux.ts (hunk 2 above)
Comment (fisher-alice) [reply apps/src/aichat/redux/aichatRedux.ts:47]: This makes sense! I'll add a TODO comment here if work is picked up later on.

File: apps/src/aichat/redux/aichatRedux.ts (hunk 3 above)
Comment (fisher-alice) [reply apps/src/aichat/redux/aichatRedux.ts:129]: Sure! Just added.

File: apps/src/aichat/redux/aichatRedux.ts [hunk 5]
```diff
@@ -42,14 +108,37 @@ const aichatSlice = createSlice({
...
+      action: PayloadAction<{id: number; status: Status}>
+    ) => {
+      const {id, status} = action.payload;
+      const chatMessage = state.chatMessages.find(msg => msg.id === id);
+      if (chatMessage) {
+        chatMessage.status = status;
+      }
+    },
+  },
+  extraReducers: builder => {
```
Comment (bencodeorg) [inline apps/src/aichat/redux/aichatRedux.ts:122]: TIL about extraReducers! Is the idea here that `isWaitingForChatResponse` would never be set explicitly by us (ie, only is a result of the state of our request to openAI), so we don't create a reducer for it?

File: apps/src/aichat/views/ChatWorkspace.tsx [hunk 6]
```diff
@@ -27,84 +24,58 @@ const ChatWorkspace: React.FunctionComponent = () => {
...
-    if (chatApiResponse.assistantResponse) {
-      const assistantChatMessage: ChatCompletionMessage = {
-        id: chatApiResponse.id + 1,
-        role: Role.ASSISTANT,
-        status: Status.OK,
-        chatMessageText: chatApiResponse.assistantResponse,
-      };
-      dispatch(addChatMessage(assistantChatMessage));
+  const showWaitingAnimation = () => {
+    if (isWaitingForChatResponse) {
```
Comment (bencodeorg) [inline apps/src/aichat/views/ChatWorkspace.tsx:43]: I originally had the waiting animation set up this way (ie, hide/show using redux), but the gif showed up intermittently and @sanchitmalhotra126 and I couldn't figure out why. We switched to a fully css based hide/show, which worked, but looks like you removed.

Just noting that I just pulled this branch and can't repro the issue we were seeing, so I think this works? 🤷

**Review by bencodeorg — APPROVED** [review]
Nice work! Learned a lot reading through this.

File: apps/src/aichat/views/ChatWorkspace.tsx (hunk 6 above)
Comment (fisher-alice) [reply apps/src/aichat/views/ChatWorkspace.tsx:43]: Weird! I'm glad you checked to make sure it's working as expected - thanks!

File: apps/src/aichat/redux/aichatRedux.ts (hunk 5 above)
Comment (fisher-alice) [reply apps/src/aichat/redux/aichatRedux.ts:122]: Yes - that's my understanding. We can leverage the state of the response from the async thunk `submitChatMessage` whose callback function returns a promise - another advantage of using a thunk!

**Pull request merged**